src/audio/kpb.c

```diff
@@ -119,11 +123,212 @@ static void kpb_free(struct comp_dev *dev)
+
+	cd->his_buf_hp.sta_addr = rballoc(RZONE_RUNTIME, SOF_MEM_CAPS_RAM,
+					  MAX_BUFFER_SIZE);
+
+	if (cd->his_buf_hp.sta_addr) {
+		trace_kpb_error("Failed to allocate space for "
+				"KPB bufefrs");
+	return -EINVAL;
+	}
+
+	cd->his_buf_hp.end_addr = cd->his_buf_hp.sta_addr + MAX_BUFFER_SIZE;
+
+#else
+#error "Wrong number of key phrase buffers configured"
+#endif
+
+	/* TODO: zeroes both buffers */
+
+	/* Initialize clients data */
+	for (i = 0; i < MAX_NO_OF_CLIENTS; i++)
+		cd->clients[i].state = KPB_CLIENT_UNREGISTERED;
+
+	return ret;
+}
+
+/**
+ * kpb_copy() - copy real time input stream into sink buffer,
+ * and in the same time buffers that input for
+ * later use by some of clients.
+ *
+ * @arg1:  kpb component.
+ *
+ * Return: integer representing either 0 - success
+ * or -EINVAL - failure.
+ */
+static int kpb_copy(struct comp_dev *dev)
+{
+	int ret;
+	struct comp_data *kpb = comp_get_drvdata(dev);
+	struct comp_buffer *source;
+	struct comp_buffer *sink;
+
+	trace_kpb("kpb_copy()");
+
+	/* Get source and sink buffers */
+	source = list_first_item(&dev->bsource_list, struct comp_buffer,
+				 source_list);
+	sink = list_first_item(&dev->bsink_list, struct comp_buffer,
+			       sink_list);
+	/* Process source data */
+	/* Check if source have data and valid pointer to read from */
+	if (source->avail > 0 && source && sink && source->r_ptr) {
+		/* Real time copying */
+		if (sink->free >= source->avail && sink->w_ptr) {
+			/* Sink and source are both ready and have space */
+			memcpy(sink->w_ptr, source->r_ptr,
+			       source->avail);
+
+			/* update source & snik data*/
+			comp_update_buffer_produce(sink, source->avail);
+			comp_update_buffer_consume(source, source->avail);
+		} else {
+			;/* What should we do if sink isn't ready(state/size)?
+			  * Simple skip is OK?
+			  */
+		}
+		/* Buffer source data internally for future use by clients */
+		if (source->avail <= MAX_BUFFER_SIZE)
+			kpb_buffer_data(kpb, source);
+	} else
+		ret = -EINVAL;
+
+	return ret;
+}
+
+static void kpb_buffer_data(struct comp_data *kpb, struct comp_buffer *source)
+{
+}
+
+/**
+ * kpb_cmd() - kpb event handler.
+ * @arg1:  kpb component.
+ * @arg2:  command to be executed
+ * @arg3:  pointer to data related with request
+ * @arg4:  maxiumum size of data
+ *
+ *
+ * Return: integer representing either 0 - success
+ * or -EINVAL - failure.
+ */
+static int kpb_cmd(struct comp_dev *dev, int cmd, void *data,
+		   int max_data_size)
+{
+	int ret = 0;
+	struct sof_ipc_ctrl_data *cdata = data;
+	struct kpb_client *cli = (struct kpb_client *)cdata->data->data;
+	uint8_t client_id = cli->id;
+
+	trace_kpb("kpb_cmd()");
+
+	switch (cmd) {
+	case KPB_EVENT_REGISTER_CLIENT:
+		ret = kpb_register_client(dev, cli);
+		break;
+	case KPB_EVENT_UNREGISTER_CLIENT:
+		/*TODO*/
+		ret = -EINVAL;
+		break;
+	case KPB_EVENT_BEGIN_DRAINNING:
+		ret = kpb_begin_drainning(dev, client_id);
+		break;
+	case KPB_EVENT_STOP_DRAINNING:
```

Comment:
@mrajwa what you might really need here is a cmd callback that can determine which sink to activate. So for the simplest case of 2 sinks where one if the keyword_detect comp and the other is the host capture PCM comp, you just need a command which tells you which sink buffer to send the data to. By default you'd send the data to the keyword detect comp buffer. When the keyword_detect detects the keyphrase, it can send a command to KPB indicating that. This can be handled with a set_value command with a 1 indicating a successful detection.